### PR TITLE
[Grid] New components

### LINKED ISSUE #148: [Grid] new component
As a consumer of Mineral I would like a grid component to layout pages and content in our application.

#### Design
https://ca.invisionapp.com/d/main#/console/11070131/250301138/preview

#### AC: 
1. Build a component / write a guideline page
1. Responsiveness
1. Show your work

#### Thoughts/Ideas:
- Grid Container (not needed)
- Grid Row
- Grid Item

#### Follow Up / Addition:
- Apply Grid to Site
- Apply Grid to first targeted product Layout as some example

#### Other Considerations:
1. Multiple Components
3. Grid Layout (probably not CSS Grid, sorry Rachel) -- why not, not supported in ie11
4. Left/Right Support -- rtl??
5. Distribution of Space
6. RTL Support (start/end)

### PR DESCRIPTION
**Pull request by kylegach**
### Description

- Add Grid & GridItem components
- Add demos & examples
- Add specs and snapshots
- Refactor Flex to share some things with Grid & GridItem
- Update Flex's gutter example to match Grid's and update Box's intro to include Grid & GridItem
- Fix display issue with Layout examples and update snapshots

### Motivation and context

Closes #148

### Screenshots, videos, or demo, if appropriate

https://148-grid--mineral-ui.netlify.com/components/grid

### How to test

1. `cd mineral-ui && npm start`
2. Navigate to [Grid](https://148-grid--mineral-ui.netlify.com/components/grid) & [GridItem](https://148-grid--mineral-ui.netlify.com/components/grid-item)
3. Confirm all looks well

### Types of changes

- New feature (non-breaking change which adds functionality)

### Checklist

* [x] Renders and functions properly in [all supported browsers](https://github.com/mineral-ui/mineral-ui#browser-support)
* [x] Automated tests written and passing
* [x] [Accessibility](http://webaim.org/intro) and [inclusivity](https://24ways.org/2016/what-the-heck-is-inclusive-design/) considered
* [x] Rendering performance (initial load time & 60fps) and [perceived performance](http://blog.teamtreehouse.com/perceived-performance) considered
* [x] Documentation created or updated
* [x] Tested in [sandbox](https://github.com/facebookincubator/create-react-app#getting-started) if new component or breaking change

### How does this PR make you feel?

Tron grid and lightcycles

### CODE REVIEW

**Comment by asagray** [comment]
This looks awesome – very easy to understand. Do you think it's appropriate to indicate under `Gutters` or anywhere else where pixel-based measurement is implied , that Mineral specifies values divisible by 2s. Or to indicate usage of tokens to set spacing?

**Comment by kylegach** [comment]
Thanks, @asagray. The gutters example shows how you can use values like `'md'`, which correspond to a spacing theme variable, `'0.5em'`, which are used directly, and `20`, which are appended with `px`.

I think the fact that the spacing theme variables are based on an 8px grid and that they're all divisible by 2 is an implementation detail, and not germane to this part of the documentation. Anyone interested in the values of the theme variables can find them [on the Theming page](http://mineral-ui.com/theming#common-scenarios-theme-structure).

File: src/website/app/demos/Grid/examples/GridItem/responsive.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
+/* @flow */
+import GridItem from '../../components/GridItem';
+import Grid from '../../components/Grid';
+import responsiveInstructions from '../../../Box/components/responsiveInstructions';
+
+export default {
+  id: 'responsive',
+  title: 'Responsive',
+  description: `TODO Many of the properties for GridItem can be responsive:
```
Comment (VictoriaVasys) [inline src/website/app/demos/Grid/examples/GridItem/responsive.js]: Still have a TODO here fyi

File: src/website/app/demos/Grid/index.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
+/* @flow */
+import React from 'react';
+import bestPractices from './bestPractices';
+
+import gridExamples from './examples/Grid';
+import gridItemExamples from './examples/GridItem';
```
Comment (VictoriaVasys) [inline src/website/app/demos/Grid/index.js]: Not sure if this was an intentional division, but there's an extra return here

File: src/website/app/demos/routes.js [hunk 3]
```diff
@@ -131,6 +131,16 @@ export default [
...
+    {
+      description: 'Grid TODO',
+      slug: 'grid',
+      title: 'Grid'
+    },
+    {
+      description: 'GridItem TODO',
+      slug: 'grid-item',
+      title: 'GridItem'
+    },
```
Comment (VictoriaVasys) [inline src/website/app/demos/routes.js]: Still some TODOs here

File: src/website/app/demos/Grid/examples/Grid/columns.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
+/* @flow */
+import GridItem from '../../components/GridItem';
+import Grid from '../../components/Grid';
+
+export default {
+  id: 'columns',
+  title: 'Columns',
+  description: `The number of columns used by Grid can be adjusted. Also see
+[GridItem's \`span\` prop](/components/grid-item#span).`,
```
Comment (VictoriaVasys) [inline src/website/app/demos/Grid/examples/Grid/columns.js]: Nice addition for connectivity.

File: src/website/app/demos/Grid/examples/Grid/gutters.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+          <Grid gutterWidth={value} key={index}>
+            <GridItem>A</GridItem>
+            <GridItem>B</GridItem>
+            <GridItem>C</GridItem>
+          </Grid>
+        ));
+
+      return <DemoLayout>{renderGrides()}</DemoLayout>;
+    }`
+};
```
Comment (VictoriaVasys) [inline src/website/app/demos/Grid/examples/Grid/gutters.js]: This example might be more visually appealing if the rendered gutter widths were ordered from least to greatest or greatest to least.

File: src/website/app/demos/Grid/bestPractices.js [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    {
+      type: 'dont',
+      description: `Don't display content directly inside GridItem. Wrap
+GridItem around components instead.`,
+      example: (
+        <Grid>
+          <GridItem>1: Shipping Info</GridItem>
+          <GridItem>2: Billing Info</GridItem>
+          <GridItem>3: Confirm Order</GridItem>
+        </Grid>
```
Comment (VictoriaVasys) [inline src/website/app/demos/Grid/bestPractices.js]: This example is a little hard to see on the site; not sure how to improve that though.

File: src/website/app/demos/Grid/examples/GridItem/gridItem.js [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+export default {
+  id: 'basic',
+  title: 'Basic Usage',
+  description: `Use [Grid](/components/grid) and GridItem to lay out
+components in a columnar fashion. With no other props applied, GridItem will
+fill the remaining space in a Grid equally with other GridItems.
+
+_Note that the last row below, in which 5 GridItems are placed in a Grid, does
+not line up with the others, because 12, the default number of columns, is not
+divisible by 5._`,
```
[the remark this answers is not in the document]
Comment (kylegach) [reply src/website/app/demos/Grid/examples/GridItem/gridItem.js]: I added a reference row showing all 12 columns to demonstrate the alignment or lack thereof. Thanks for the feedback.

File: src/website/app/demos/Grid/bestPractices.js (hunk 6 above)
Comment (kylegach) [reply src/website/app/demos/Grid/bestPractices.js]: We kinda _want_ it to look bad.

File: src/website/app/demos/Grid/examples/Grid/gutters.js (hunk 5 above)
Comment (kylegach) [reply src/website/app/demos/Grid/examples/Grid/gutters.js]: Also updated Flex's gutter example.

**Comment by kylegach** [comment]
All necessary comments have been addressed.

**Comment by kylegach** [comment]
@trotzig — Getting this error on the [happo job](https://travis-ci.org/mineral-ui/mineral-ui/jobs/368241859) (tried multiple times):

```
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
```

From the log, it appears that the Chrome snaps work, but it stalls on IE.

Any ideas?

**Comment by trotzig** [comment]
Okay, it looks like a combination of things: 
- The internet explorer job queue was backed up (still is, but things should be better soon). 
- Some of the ie worker nodes had gone offline (bringing them back up now)

Thanks for letting me know! You should be able to retrigger the job and it should finish in time. 

If you want to optimize for build speed, you could consider turning off the internet explorer target for happo. Your chrome jobs finish in less than a minute, and I think IE takes at least three times as long.  

@kylegach

**Comment by kylegach** [comment]
That worked. Thanks, Henric!

Good call on re-considering our use of the IE diffs. Will discuss with the team.

File: src/library/index.js [hunk 8]
```diff
@@ -17,6 +17,7 @@ export { default as Checkbox, CheckboxGroup } from './Checkbox';
 export { default as Dropdown } from './Dropdown';
 export { default as Flex, FlexItem } from './Flex';
 export { FormField, FormFieldset, FormFieldDivider } from './Form';
+export { default as GridRow, GridItem } from './Grid';
```
Comment (brentertz) [inline src/library/index.js]: I believe that `GridRow` should be `Grid`, yes?

File: src/website/app/pages/ComponentStatus/componentStatus.md [hunk 9]
```diff
@@ -35,48 +36,46 @@ Statuses:
...
 | [TextArea](/components/text-area)           | [available] |
-| [Text](/components/text) ~ new              | [available] |
+| [Text](/components/text)                    | [available] |
 | [TextInput](/components/text-input)         | [available] |
 | [ThemeProvider](/components/theme-provider) | [available] |
 | [Tooltip](/components/tooltip)              | [available] |
-| Grid                                        | [planned]   |
 | Toggle                                      | [planned]   |
 | Table                                       |             |
+| Alert                                       |             |
```
Comment (brentertz) [inline src/website/app/pages/ComponentStatus/componentStatus.md]: Oooh, thanks for sorting these...

File: src/library/Grid/__tests__/Grid.spec.js [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+  return shallow(
+    <Grid {...gridProps}>
+      <GridItem {...gridItemProps} />
+    </Grid>
+  );
+}
+
+describe('Grid', () => {
+  it('renders', () => {
+    const grid = shallowGrid([{}, {}]);
```
Comment (brentertz) [inline src/library/Grid/__tests__/Grid.spec.js]: While it really doesn't matter, there is no reason to pass an array here.  You could simply pass 2 props objects.

File: src/library/index.js (hunk 8 above)
Comment (kylegach) [reply src/library/index.js]: 🤦‍♂️

**Pull request merged**